core/src/main/scala/zio/jdbc/JdbcDecoder.scala

```diff
@@ -25,678 +26,504 @@ import scala.collection.immutable.ListMap
  * A type class that describes the ability to decode  a value of type `A` from
  * a `ResultSet`.
  */
-trait JdbcDecoder[+A] {
-  def unsafeDecode(rs: ResultSet): A
+trait JdbcDecoder[+A] { self =>
+  def unsafeDecode(rs: RowState): (RowState, A)
```

Comment:
Would be a bit more performant to change this to:

```suggestion
  def unsafeDecode(i: Int, rs: ResultSet): (Int, A)
```

Why? Because it's only 1 allocation for the tuple, rather than 3 allocations (one for input row state, one for output row state, and one for tuple).

Reply:
👍

Reply:
We need the resultSet on the out as well. I can change it.